cli/tools/lint/plugins.rs

```diff
@@ -0,0 +1,558 @@
+    )
+  };
+
+  Ok(PluginHost {
+    worker,
+    install_plugins_fn,
+    run_plugins_for_file_fn,
+    tx: tx_res,
+    rx: rx_req,
+    logger: logger.clone(),
+  })
+}
+
+#[derive(Debug, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct PluginInfo {
+  pub name: String,
+  pub rule_names: Vec<String>,
+}
+
+impl PluginInfo {
+  pub fn get_rules(&self) -> Vec<String> {
+    let mut rules = Vec::with_capacity(self.rule_names.len());
+
+    for rule_name in &self.rule_names {
+      rules.push(format!("{}/{}", self.name, rule_name));
+    }
+
+    rules
+  }
+}
+
+impl PluginHost {
+  fn create(logger: PluginLogger) -> Result<PluginHostProxy, AnyError> {
+    let (tx_req, rx_req) = channel(10);
+    let (tx_res, rx_res) = channel(10);
+
+    logger.debug("spawning thread");
+    let logger_ = logger.clone();
+    let join_handle = std::thread::spawn(move || {
+      let logger = logger_;
+      logger.debug("PluginHost thread spawned");
+      let start = std::time::Instant::now();
+      let fut = async move {
+        let runner =
+          create_plugin_runner_inner(logger.clone(), rx_req, tx_res).await?;
+        // TODO(bartlomieju): send "host ready" message to the proxy
+        logger.debug("running host loop");
+        runner.run_loop().await?;
+        logger.debug(&format!(
+          "PluginHost thread finished, took {:?}",
+          std::time::Instant::now() - start
+        ));
+        Ok(())
+      }
+      .boxed_local();
+      tokio_util::create_and_run_current_thread(fut)
+    });
+
+    logger.debug(&format!("is thread finished {}", join_handle.is_finished()));
+    let proxy = PluginHostProxy {
+      tx: tx_req,
+      rx: Arc::new(tokio::sync::Mutex::new(rx_res)),
+      plugin_info: Arc::new(Mutex::new(vec![])),
+      join_handle,
+      logger,
+    };
+
+    Ok(proxy)
+  }
+
+  async fn run_loop(mut self) -> Result<(), AnyError> {
+    self.logger.debug("waiting for message");
+    while let Some(req) = self.rx.recv().await {
+      self.logger.debug("received message");
+      match req {
+        PluginHostRequest::LoadPlugins {
+          specifiers,
+          exclude_rules,
+        } => {
+          let r = self.load_plugins(specifiers, exclude_rules).await;
+          let _ = self.tx.send(PluginHostResponse::LoadPlugin(r)).await;
+        }
+        PluginHostRequest::Run {
+          serialized_ast,
+          file_path,
+          source_text_info,
+          maybe_token,
+        } => {
+          let start = std::time::Instant::now();
+          let r = match self.run_plugins(
+            &file_path,
+            serialized_ast,
+            source_text_info,
+            maybe_token,
+          ) {
+            Ok(()) => Ok(self.take_diagnostics()),
+            Err(err) => Err(err),
+          };
+          self.logger.debug(&format!(
+            "Running rules took {:?}",
+            std::time::Instant::now() - start
+          ));
```

Comment:
Maybe we should just use regular `log::debug!` for stuff like this so that we're not allocating strings for debug printing when debug is off? Maybe `PluginLogger` should only be used for op_print and then we use the regular `log` crate elsewhere?

Reply:
I need to double check, when I did that previously it caused hanging in the LSP when logging was on.